src/deploy/permission/index.ts

```diff
@@ -108,18 +108,50 @@ export const selectIsUserAnyOwner = createSelector(
   },
 );
 
-export const selectPermsByAccount = createSelector(
+/**
+ * Nested map: Role -> Env -> Permissions scoped by role and env
+ * {
+ *  [roleId]: {
+ *    [envId]: Permission[];
+ *  };
+ * }
+ */
+export const selectRoleToEnvToPermsMap = createSelector(
+  schema.roles.selectTableAsList,
+  schema.environments.selectTableAsList,
+  selectPermissionsAsList,
+  (roles, envs, perms) => {
+    const mapper: { [key: string]: { [key: string]: Permission[] } } = {};
+    for (const role of roles) {
+      const envToPermMap: { [key: string]: Permission[] } = {};
+      for (const env of envs) {
+        const envPerms = perms.filter(
+          (perm) => perm.roleId === role.id && perm.environmentId === env.id,
+        );
+        if (envPerms.length > 0) {
+          envToPermMap[env.id] = envPerms;
+        } else {
+          envToPermMap[env.id] = [];
+        }
```

Comment:
Won't `.filter` return an empty array if none of the entries satisfy the filter condition?

```suggestion
        envToPermMap[env.id] = perms.filter(
          (perm) => perm.roleId === role.id && perm.environmentId === env.id,
        );
```

Reply:
good catch! updated